File: algorithms/medium/distinct_prime_factors_of_product_of_array.ex
defmodule Solution do
  @spec distinct_prime_factors(nums :: [integer]) :: integer
  def distinct_prime_factors(nums) do
    nums
    |> Enum.reduce(%MapSet{}, fn n, st ->
      prime_factors(2, n, st)
    end)
    |> MapSet.size()
  end

  def prime_factors(p, n, st) when p > n, do: st

  def prime_factors(p, n, st) do
    if rem(n, p) === 0 do
      prime_factors(p, div(n, p), MapSet.put(st, p))
    else
      prime_factors(p + 1, n, st)
    end
  end
end

File: algorithms/medium/distinct_prime_factors_of_product_of_array.cpp
class Solution {
public:
  unordered_set<int> st;
  int distinctPrimeFactors(vector<int> &nums) {
    for (auto n : nums) {
      primeFactor(n);
    }
    return st.size();
  }
  void primeFactor(int n) {
    int p = 2;
    while (n >= p) {
      if (n % p == 0) {
        st.insert(p);
        n /= p;
      } else {
        p++;
      }
    }
  }
};
